### DISCUSSION TITLE
Simple guide on setting up auto updater

### DISCUSSION THREAD

**Discussion by Ripwords** [tauri-apps/tauri#2776] [Q&A] [answered]
Hi, I'm new to Tauri and i have been developing a simple app using Tauri for the past few days. 

I wanted to configure an auto updater similar to how electron does it. After reading the docs and if I am not mistaken, am i correct to assume i would need a backend server that is able to respond with the proper JSON file in order for the updater to work?

Prior to using Tauri, I am using Electron-Builder and Github Releases to configure auto updates for my applications. As I am not planning to implement a dedicated release server anytime soon, may i know are there any methods i could configure Tauri to get updater working with just Github Releases?

**Comment by amrbashir** [comment, marked as answer]
you can just add a json in a github gist and point the updater to it.

**Reply by Ripwords** [reply]
Ah i didn't know i could do that, thanks i will give it a go. But since I'm not exactly sure how to do that yet, do you have any examples that i can check out? Thanks in advance

**Reply by FabianLars** [reply]
@Ripwords If you want to use gh gist:
Create a gist with a json file similar to this one: https://tauri.studio/en/docs/guides/updater/#update-file-json-format and point the updater to a url like "https://gist.githubusercontent.com/USERNAME/GIST_ID/raw"

**Reply by Ripwords** [reply]
Thanks for the tip I will look into it

**Reply by Ripwords** [reply]
Just got it working, can't believe the process is so painless. Thanks again

**Reply by dsegovia90** [reply]
The first link that @FabianLars commented is broken. I'm pretty sure this format is what you're refering too:
https://tauri.studio/docs/distribution/updater#update-file-json-format

**Reply by FabianLars** [reply]
yep, that's it (it's the same headline after all 😅), the structure of our docs changed again since i last edited that link and i didn't update all the references in issues&discussions yet because i have a feeling it will change again lol

**Reply by wolfgangmeyers** [reply]
I think it moved again 😄 https://tauri.studio/v1/guides/distribution/updater

**Reply by sinister296** [reply]
How do you exactly point the updater to the github gists?

**Reply by HuakunShen** [reply]
For example, this is my gist https://gist.github.com/HuakunShen/c71fb4093e68aa2ed24860a5ea812708

Click on `raw` button to get the raw file, it looks like this `https://[gist.githubusercontent.com/HuakunShen/c71fb4093e68aa2ed24860a5ea812708/raw/cabf1f1eb131ffdad7f03d3d31bcacedd7a0768a/ezup-update.json`

You need to remove the last two parts which will be updated once you update the gist.

So the result will be ``https://[gist.githubusercontent.com/HuakunShen/c71fb4093e68aa2ed24860a5ea812708/raw/`

Add it to the the list of `endpoints` in `tauri.conf.json`.

**Reply by mostafa-fallaha** [reply]
can someone explain to me what should i put here? 
"url": "https://github.com/username/reponame/releases/download/v1.0.0/app-x64-setup.nsis.zip" 

because i can't find the "releases/download/v1.0.0/app-x64-setup.nsis.zip" in my repo, and i didnt push the target folder in src-tauri

thanks in advance.

**Reply by CalvinWilkinson** [reply]
@mostafa-fallaha 

The link should be the same as if somebody were to go to a release asset and then click on it. 

![image](https://github.com/tauri-apps/tauri/assets/85414302/79a0f693-ed06-4d8d-84e1-a95c062a7b81)

**Reply by mostafa-fallaha** [reply]
thanks a lot man, appreciate it.

**Reply by HoaX7** [reply]
@HuakunShen can you please explain how to connect the updater to the gist? i pasted the raw link like you mentioned but that isnt auto updating even when i edit the latest.json

**Reply by HoaX7** [reply]
and i used .exe to like install the app

**Reply by CalvinWilkinson** [reply]
@HoaX7 

Just go to the gist and get the URL.  Just add that URL to the endpoint array in your _**tauri.conf.json**_ file.

This URL:
<img width="1673" height="455" alt="image" src="https://github.com/user-attachments/assets/2c3d694a-cf03-478a-a5e7-b78ddbf5b3fc" />


Add to this array:
```json
"plugins": {
    "updater": {
        "pubkey": "",
        "endpoints": [
            "https://gist.github.com/CalvinWilkinson/abcdabcdabcdabcdabcdabcdabcdab"  👈🏻 Add here
        ]
    }
},
```

> [!NOTE]
> The GIST must be public.  A private GIST will not work unless you have an update server taking the request and serving the GIST data back to the application.

**Comment by jecsham** [comment]
Guide moved to: https://tauri.studio/v1/guides/distribution/updater

**Reply by jecsham** [reply]
Moved again to https://tauri.app/v1/guides/distribution/updater/

**Comment by holmesadrianh** [comment]
Hi guys, is there a way to automatically generate the json file when the apps are built? When I do a tauri build and have bundles available, I would like to streamline the process of creating the update json file to then upload to my server. Currently I need to copy paste the signature found in the sig files to the json file manually. It would be great to have this done automatically. Then with an automatic upload to the server the build process would be really smooth. Unless there is already a build process or workflow that I am not aware of that takes care of this?

**Reply by PhiloNL** [reply]
If you want to try an automated process, I've been working on adding support for [Tauri at Unlock](https://unlock.sh/tauri-applications). It's now in beta; you use GitHub actions to generate the build artifacts and publish your release. Unlock will take care of everything else, like generating an updater endpoint with the binaries and signatures.

**Reply by mitch1009** [reply]
> Hi guys, is there a way to automatically generate the json file when the apps are built? When I do a tauri build and have bundles available, I would like to streamline the process of creating the update json file to then upload to my server. Currently I need to copy paste the signature found in the sig files to the json file manually. It would be great to have this done automatically. Then with an automatic upload to the server the build process would be really smooth. Unless there is already a build process or workflow that I am not aware of that takes care of this?

This weekend i decided to embark on this. i managed to craft a solution. i will have to test it thoroughly then i can make it available for the community

**Reply by anth0ni3** [reply]
@mitch1009 Could you share your fix?

**Reply by mitch1009** [reply]
It crumbled miserably, but I will fix it and share the solution.

**Comment by andreial** [comment]
Hi, is there a way to get the auto updater work with private GitHub repos?

**Reply by jecsham** [reply]
You just need to be sure your  `.json` file is served publicly

**Reply by andreial** [reply]
Right, but the `.json` file urls would be pointing to files contained in the release, which is private and inaccessible unless you also provide a GitHub auth token. Or am I missing something?

**Reply by mooijtech** [reply]
@andreial You can create a separate public repository only containing your releases by duplicating the releases from your private repository: https://github.com/softprops/action-gh-release
In the same public repository you can keep track of your `.json` file and make Tauri point to it.

**Reply by CalvinWilkinson** [reply]
@mooijtech 
I have seen the same solution in other places but the release asses in the public repo is still public.  I think that the thing people are trying to figure out is to NOT have the release assets public.  

For example, I have a client right now who does not want the application assets public, which would be public if the repo and release were public.

**Reply by VarshilJPatel** [reply]
@andreial use github person access token for Accessing private repositories.

**Comment by HuakunShen** [comment]
Do we have to implement a update server? Is there a template we can use?

I am looking at https://github.com/vercel/hazel, which should work for [electron](https://www.electronjs.org/docs/latest/tutorial/updates#step-1-deploying-an-update-server)

**Reply by ku9nov** [reply]
Hi 
You can check out [faynoSync](https://github.com/ku9nov/faynoSync)
 — an update server that supports Tauri/Electron, with built-in app statistics and flexible configuration. Would love your feedback.

**Comment by WhyNeet** [comment]
`Error Key generation aborted: Unable to find the private key`
I am having this issue when signing a file with `cargo tauri signer sign "PATH_TO_ZIP_RELEASE"`
Can anybody help me with this please?

**Reply by vlev1n** [reply]
I was stuck on this problem until I found `pnpm tauri signer -h` and `pnpm tauri signer sign -h` is different. You may try `pnpm tauri signer sign -k your_private_key -p your_private_key_password path_of_your_zip_or_tgz_file`. `pnpm tauri signer sign -h` tells you what `-k` and `-p` is.
Output of `pnpm tauri signer sign -h`
```
Sign a file

Usage: pnpm run tauri signer sign [OPTIONS] <FILE>

Arguments:
  <FILE>  Sign the specified file

Options:
  -k, --private-key <PRIVATE_KEY>            Load the private key from a string
  -v, --verbose...                           Enables verbose logging
  -f, --private-key-path <PRIVATE_KEY_PATH>  Load the private key from a file
  -p, --password <PASSWORD>                  Set private key password when signing
  -h, --help                                 Print help
  -V, --version                              Print version
```
Output of `pnpm tauri signer -h`
```
Tauri updater signer

Usage: pnpm run tauri signer [OPTIONS] <COMMAND>

Commands:
  sign      Sign a file
  generate  Generate keypair to sign files
  help      Print this message or the help of the given subcommand(s)

Options:
  -v, --verbose...  Enables verbose logging
  -h, --help        Print help
  -V, --version     Print version
```
Better use single quote if your password contains `!` mark.
If it works as expected, a string of signature will be returned.

**Comment by cage1618** [comment]
How to enable automatic updates for applications installed through both MSI and NSIS installation packages?
My endpoints use static json file, but in the 'platforms' field, only one of MSI or NSIS can be specified. Is there a way to support both MSI and NSIS?

**Reply by FabianLars** [reply]
Not directly, you'll need 2 latest.json files which also means you need to build the app twice (one for each endpoint url).

**Reply by cage1618** [reply]
got it, thanks

**Comment by delebash** [comment]
First thanks for Tauri it is so much smaller deployment than electron.  I have my deployment to github and updater working but I have to say it is a lot of work just to make a new release using the latest.json manual not github actions method.  So, it looks like every time I want to publish a new release I have to touch 5 files.  I have to change the version number in tauri.conf.json, Cargo.toml, package.json, and latest.json with the addition of opening the new key files and pasting that into latest.json and pasting the new download paths.  With electron builder I could setup my config once and then all I had to do was change the version in the package.json to create a new release and upload exe files.  A simple npm run publish is all that was needed after the initial config. Unless I am doing this wrong I think the deployment and update procedure needs to be looked at.  Thnaks!

**Reply by FabianLars** [reply]
>  I have to change the version number in tauri.conf.json, Cargo.toml, package.json

Tauri itself only cares about one of those. If you set a version in tauri.conf.json it will use that. If you set that version field to a file path to your package.json file it will use that. If you remove the version from tauri.conf.json it will fall back to cargo.toml.

The version fields in tauri.conf.json are optional (can be completely removed), only cargo requires a version in cargo.toml (can be left to 0.0.0 or whatever if you set one of the other versions).

The rest of your message about latest.json and signature files is of course correct. We may add some automation into the tauri cli or an external tool at some point but there are currently no plans.

To have some more data for a potential feature, what service/provider are you using to host your update? Assuming you're not using tauri-action because you're not using github releases for that.

**Reply by CalvinWilkinson** [reply]
You can automate it.  Though I am using a custom proxy server for updates for my client, once the proxy server is up and running, the update of 3 pieces of info in the _**tauri.conf.json**_ file are all automated during the release process.  I do all of this using deno and it works great.

**Reply by brightmadueke** [reply]
please how can i use github actions for my updater. i am new here

**Comment by delebash** [comment]
Thanks for letting me know.  I was looking for that info in the docs but could not find it.  I then came across an article on how to do it and they said you needed to change all three.  I am using github and I just got tauri-action working in a manual release mode.  Thanks, Cheers!
